KAFKA-17578: Remove partitionRacks from TopicMetadata

FrankYang0529 opened the pull request:
The ModernGroup#subscribedTopicMetadata takes too much memory due to `partitionRacks`. This is not being used at the moment as the consumer protocol does not support rack aware assignments.

A heap dump from a group with 500 members, 2K subscribed topic partitions shows 654,400 bytes used for partitionRacks. The rest of the ConsumerGroup object holds 822,860 bytes. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dajac commented on group-coordinator/group-coordinator-api/src/main/java/org/apache/kafka/coordinator/group/api/assignor/SubscribedTopicDescriber.java:
```diff
@@ -38,14 +36,4 @@ public interface SubscribedTopicDescriber {
-     * @return The set of racks corresponding to the replicas of the topic's partition.
-     *         If the topic Id does not exist, an empty set is returned.
-     */
-    Set<String> racksForPartition(Uuid topicId, int partition);
```
We cannot remove this from the public interface. Let’s return an empty set for now.

dajac commented on group-coordinator/src/main/resources/common/message/ConsumerGroupPartitionMetadataValue.json:
```diff
@@ -27,14 +27,7 @@
       { "name": "TopicName", "versions": "0+", "type": "string",
         "about": "The topic name." },
       { "name": "NumPartitions", "versions": "0+", "type": "int32",
-        "about": "The number of partitions of the topic." },
```
Same. Unfortunately, we cannot remove fields because we released it. We must keep it. Let’s put a comment explaining that we don’t use it.

dajac reviewed (commented):
@FrankYang0529 Thanks for working on this.  I left a few high level comments to start with.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicMetadata.java:
```diff
@@ -130,45 +107,24 @@ public String toString() {
     }
 
     public static TopicMetadata fromRecord(
             ShareGroupPartitionMetadataValue.TopicMetadata record
```
nit: Not related to your changes but the indentation is incorrect here.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicMetadata.java:
```diff
@@ -130,45 +107,24 @@ public String toString() {
         return new TopicMetadata(
                 record.topicId(),
                 record.topicName(),
-                record.numPartitions(),
-                partitionRacks);
+                record.numPartitions());
```
nit: ditto.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/Assertions.java:
```diff
@@ -242,27 +242,6 @@ private static void assertApiMessageAndVersionEquals(
                 assertEquals(expectedTopicMetadata.topicId(), actualTopicMetadata.topicId());
                 assertEquals(expectedTopicMetadata.topicName(), actualTopicMetadata.topicName());
                 assertEquals(expectedTopicMetadata.numPartitions(), actualTopicMetadata.numPartitions());
-
```
I suppose that we can keep this code if we keep the fields in the record.

FrankYang0529 commented:
Hi @dajac, thanks for review and suggestion. I addressed all comments. However, both `GroupMetadataManagerTest#testGroupEpochBumpWhenNewStaticMemberJoins` and `GroupMetadataManagerTest#testNewJoiningMemberTriggersNewTargetAssignment` are failed cases. I will try to fix them this weekend.

FrankYang0529 commented:
Hi @dajac, I fixed both `testGroupEpochBumpWhenNewStaticMemberJoins` and `testNewJoiningMemberTriggersNewTargetAssignment`. Originally, different `partitionRacks` triggers subscription metadata update which generates a new consumer group subscription metadata record. Since we remove the field, we have to remove the related record in test cases.

dajac commented:
@FrankYang0529 There are a few conflicts. Could you please address them?

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/ModernGroup.java:
```diff
@@ -389,25 +388,10 @@ public Map<String, TopicMetadata> computeSubscriptionMetadata(
                     topicImage.name(),
-                    topicImage.partitions().size(),
-                    partitionRacks)
+                    topicImage.partitions().size())
```
nit: Could you please put the last closing parenthesis on the next line?

```
 newSubscriptionMetadata.put(topicName, new TopicMetadata(
    topicImage.id(),
    topicImage.name(),
    topicImage.partitions().size()
));
```

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicMetadata.java:
```diff
@@ -130,45 +107,24 @@ public String toString() {
             record.topicName(),
-            record.numPartitions(),
-            partitionRacks);
+            record.numPartitions());
```
nit: Ditto about the closing parenthesis.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicMetadata.java:
```diff
@@ -130,45 +107,24 @@ public String toString() {
-                partitionRacks);
+            record.topicId(),
+            record.topicName(),
+            record.numPartitions());
```
nit: ditto.

dajac reviewed (commented):
@FrankYang0529 Thanks for the update. I left a few nits for consideration.

FrankYang0529 committed:
KAFKA-17578: Remove partitionRacks from TopicMetadata

Signed-off-by: PoAn Yang <payang@apache.org>

FrankYang0529 commented:
Hi @dajac, thanks for the review. Updated it.

dajac reviewed (approved):
LGTM, thanks.

dajac commented:
@FrankYang0529 Thanks for the patch. I just merged it. We need to work on another strategy to trigger rebalances based on rack changes. I have a few ideas. If you are interested, you could pick it up.

FrankYang0529 commented:
Hi @dajac, I'm glad to handle it. Feel free to assign to me. Thank you 😄

dajac commented:
@FrankYang0529 I filed https://issues.apache.org/jira/browse/KAFKA-17747.

FrankYang0529 commented:
> @FrankYang0529 I filed https://issues.apache.org/jira/browse/KAFKA-17747.

Thank you. May I try it?
